eslint-plugin-expensify/tests/no-api-side-effect-actions.test.js

```diff
@@ -0,0 +1,71 @@
+const RuleTester = require('eslint').RuleTester;
+const rule = require('../no-api-side-effect-actions');
+const message = require('../CONST').MESSAGE.NO_API_SIDE_EFFECTS_ACTIONS;
+
+const ruleTester = new RuleTester({
+    parserOptions: {
+        ecmaVersion: 6,
+        sourceType: 'module',
+    },
+});
+
+ruleTester.run('no-api-side-effect-actions', rule, {
+    valid: [
+        {
+            code: `
+                function test() {
+                    API.write('Report_AddComment', params);
+                }
+            `,
+            filename: './src/libs/actions/Test.js',
+        },
+        {
+            code: `
+                import Action from './src/libs/actions/Action.js'
+                function test() {
+                    Action.perform(params);
+                }
+            `,
+            filename: './src/libs/actions/Test.js',
+        },
+        {
+            // Test that two functions in one file with different API/action calls are valid
+            code: `
+                import Action from './src/libs/actions/Action.js'
+                function test() {
+                    API.write('Report_AddComment', params);
+                }
+                function test2() {
+                    Action.perform(params);
+                }
+            `,
+            filename: './src/libs/actions/Test.js',
+        },
+        {
+            // Make sure that we will not test a file which is not in '/actions/'
+            code: `
+                import Action from './src/libs/actions/Action.js'
+                function test() {
+                    API.write('Report_AddComment', params);
+                    Action.perform(params);
+                }
+            `,
+            filename: './src/libs/notActions/Test.js',
+        },
+    ],
+    invalid: [
+        {
+            code: `
+                import Action from './src/libs/actions/Action.js'
+                function test() {
+                    API.write('Report_AddComment', params);
+                    Action.perform(params);
```

Comment:
I _think_ the Action would have another API call inside (or another action but eventually with API call in the end) so technically this would be two API calls in one function/method which is the camouflaged `no-multiple-api-calls` rule.

Reply:
Ok making more sense now, thanks! So the assumption is that `Action.something()` has an `API.write()` or `API.read()` in it because it (1) is a function and (2) it is exported from the `/actions` directory. 

Do we check this or guarantee that the function is also calling `API.write()` or `API.read()` or is that just implied by 1 and 2 ? 

Seems like there are actions that do not call API at all. Should we let them be called or not? How about things that set Onyx data? Should those be allowed to be called?

I think the change is fine just pushing to get the guidance a bit clearer.

Reply:
@marcaaron for an example you could look at `validateLogin` here: https://github.com/Expensify/App/blob/0b5f84330063763de41896d5fe74eacf432fa9c3/src/libs/actions/User.js#L222

We validate the login and then call the action to get UserDetails, which goes ahead and calls another API command. This should instead be done in one single API command.

Reply:
@yuwenmemon Ok, so a good "invalid" example of chaining would be something like:

```js
function validateAndGetDetails(params) {
    API.makeRequestWithSideEffects('ValidateEmail', params)
        .then((response) => {
            API.write('GetUserDetails', response);
        });
}
```

As we can't chain with `write` or `read` yea?

Reply:
>Do we plan to make sure that each action//API call has its own js file in /actions/ in which there would be only one function and that one function is exported

At the moment an "action method" is very loose but in general is something that either:

- Sets data in Onyx
- Calls an API and also sets data in Onyx

There are unfortunate actions that do things like format data, but those are not really actions at all and just colocated around actions. All of that kind of stuff I would expect will light up with a lot of errors once this change is introduced.

Reply:
Yeah I see what you mean.

> So this rule would be more helpful if it didn't target old code (e.g. DeprecatedAPI usages are fine but API usages are not fine). Not sure if that is possible - but I'd think we want to avoid people trying to "fix" old deprecated code when those rules only apply to the refactors we are planning or new APIs being added.

I like this. I think it's possible because we can just exclude any action that is making a call to `DeprecatedAPI` right? @mountiny do you think that would be possible?

Reply:
@marcaaron @yuwenmemon That makes sense to leave `DeprecatedAPI` out.

In that case, should we also update the `no-multiple-api-calls` to only look for `token.value === 'API'`? Because now it would flag the old api too.
https://github.com/Expensify/eslint-config-expensify/blob/9a56a5124e2a5340a020283a513d591381dcf852/eslint-plugin-expensify/no-multiple-api-calls.js#L11

Reply:
Yeah, that would make sense to me :+1: